File: vol2/ch23/LeetCode55-2.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n=nums.size();
        if(n==1) return true;
        if(nums[0]==0) return false;
        int dp=nums[0];
        for(int i=1;i<n-1;i++) {
            dp=max(dp,i+nums[i]);
            if(dp>=n-1) return true; 
            if(dp==i) return false;
        }
        return true;
    }
};

File: vol2/ch23/LeetCode55-2.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n=len(nums)
        if n==1:return True
        if nums[0]==0:return False
        dp=nums[0]
        for i in range(1,n-1):
            dp=max(dp,i+nums[i])
            if dp>=n-1:return True
            if dp==i:return False
        return True
